### PR TITLE
UCT/IB: Add async events callback infra

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
     *numa_node = (status == UCS_OK) ? n : -1;
 }
 
+static void
+uct_ib_device_async_event_dispatch(uct_ib_device_t *dev,
+                                   uct_ib_async_event_t *event)
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:200]: const*

File: src/uct/ib/base/ib_device.c [hunk 2]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+{
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+    int ret;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_put(uct_ib_async_event, &dev->async_events_hash, *event, &ret);
+    entry = &kh_value(&dev->async_events_hash, iter);
+    if (ret == 0) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:209]: use ucs_kh_put_t enum

File: src/uct/ib/base/ib_device.c [hunk 3]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+        if (entry->cb != NULL) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:210]: in which case we can have entry with NULL callback?

File: src/uct/ib/base/ib_device.c [hunk 4]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+            ucs_callbackq_add_safe(entry->cbq, entry->cb, entry->arg,
+                                   UCS_CALLBACKQ_FLAG_ONESHOT);
+            kh_del(uct_ib_async_event, &dev->async_events_hash, iter);
+        } else {
+            entry->flag++;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:215]: why ++ and not =1?

File: src/uct/ib/base/ib_device.c [hunk 5]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+}
+
+ucs_status_t
+uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
+                                    enum ibv_event_type event_type,
+                                    uint32_t resource_id,
+                                    ucs_callbackq_t *cbq,
+                                    ucs_callback_t cb, void *arg)
+{
+    uct_ib_async_event_t event = {};
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:231]: why need to init?

File: src/uct/ib/base/ib_device.c [hunk 6]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    uct_ib_async_event_val_t *entry;
+    ucs_status_t status = UCS_ERR_NO_MEMORY;
+    khiter_t iter;
+    int ret;
+
+    event.event_type  = event_type;
+    event.resource_id = resource_id;
+    ucs_spin_lock(&dev->async_event_lock);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:239]: space line before

File: src/uct/ib/base/ib_device.c [hunk 7]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    iter = kh_put(uct_ib_async_event, &dev->async_events_hash, event, &ret);
+    entry = &kh_value(&dev->async_events_hash, iter);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:241]: align = on column

File: src/uct/ib/base/ib_device.c [hunk 8]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    if (ret == 0) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:242]: use ucs_kh_put_t enum

File: src/uct/ib/base/ib_device.c [hunk 9]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    ucs_spin_unlock(&dev->async_event_lock);
+
+    return status;
+}
+
+void uct_ib_device_async_event_reset(uct_ib_device_t *dev,
+                                     enum ibv_event_type event_type,
+                                     uint32_t resource_id)
+{
+    uct_ib_async_event_t event = {};
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:262]: why need to init

File: src/uct/ib/base/ib_device.c [hunk 10]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+        entry->cb   = cb;
+        entry->arg  = arg;
+        status      = UCS_INPROGRESS;
+    }
+    ucs_spin_unlock(&dev->async_event_lock);
+
+    return status;
+}
+
+void uct_ib_device_async_event_reset(uct_ib_device_t *dev,
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:258]: uct_ib_device_async_event_clear

File: src/uct/ib/base/ib_device.c [hunk 11]
```diff
@@ -176,6 +195,82 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+                                     enum ibv_event_type event_type,
+                                     uint32_t resource_id)
+{
+    uct_ib_async_event_t event = {};
+    khiter_t iter;
+
+    event.event_type  = event_type;
+    event.resource_id = resource_id;
+    ucs_spin_lock(&dev->async_event_lock);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:267]: space line before

File: src/uct/ib/base/ib_device.c [hunk 12]
```diff
@@ -282,7 +378,7 @@ void uct_ib_handle_async_event(uct_ib_device_t *dev, uct_ib_async_event_t *event
     case IBV_EVENT_PORT_ERR:
         snprintf(event_info, sizeof(event_info), "%s on port %d",
                  ibv_event_type_str(event->event_type), event->port_num);
-        level = UCS_LOG_LEVEL_ERROR;
+        level = UCS_LOG_LEVEL_DIAG;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:381]: why changed lever? (for others as well)

File: src/uct/ib/base/ib_device.c [hunk 13]
```diff
@@ -457,6 +554,15 @@ void uct_ib_device_cleanup(uct_ib_device_t *dev)
 
     ucs_debug("destroying ib device %s", uct_ib_device_name(dev));
 
+    /* TODO uncomment once uct_ep_flush(CANCEL) non blocking done
+     * ucs_assert(kh_size(&dev->async_events_hash) == 0); */
+    kh_destroy_inplace(uct_ib_async_event, &dev->async_events_hash);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:559]: i'd add ucs_debug() message here if non-empty, and after flush(CACNEL) nb is done, change it to ucs_warn

File: src/uct/ib/base/ib_device.h [hunk 14]
```diff
@@ -155,6 +156,28 @@ typedef struct uct_ib_device_spec {
 
 KHASH_TYPE(uct_ib_ah, struct ibv_ah_attr, struct ibv_ah*);
 
+
+typedef struct uct_ib_async_event {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h:160]: pls add comment to struct

File: src/uct/ib/base/ib_device.h [hunk 15]
```diff
@@ -155,6 +156,28 @@ typedef struct uct_ib_device_spec {
...
+    union {
+        uint8_t         port_num;
+        uint32_t        qp_num;
+        uint32_t        dct_num;
+        void            *cookie;
+        uint32_t        resource_id;
+    };
+} uct_ib_async_event_t;
+
+typedef struct {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.h:171]: space line before
pls add comment to struct and its fields

File: test/gtest/uct/ib/test_ib_event.cc [hunk 16]
```diff
@@ -22,24 +23,33 @@ extern "C" {
 
 class uct_p2p_test_event : public uct_p2p_test {
 private:
-    void rc_mlx5_ep_to_err(entity &e, uint32_t *qp_num_p) {
+    uint32_t rc_mlx5_qp_num(entity &e) {
+#if HAVE_MLX5_HW
+        uct_rc_mlx5_ep_t *ep = (uct_rc_mlx5_ep_t *)e.ep(0);
+        return ep->tx.wq.super.qp_num;
+#else
+        return 0;
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:31]: maybe ucs_fatal()?

File: test/gtest/uct/ib/test_ib_event.cc [hunk 17]
```diff
@@ -65,38 +75,98 @@ class uct_p2p_test_event : public uct_p2p_test {
...
+        const resource *r = dynamic_cast<const resource*>(GetParam());
+
+        if (r->tl_name == "rc_mlx5") {
+            return rc_mlx5_qp_num(e);
+        } else {
+            return rc_verbs_qp_num(e);
+        }
+    }
+
+    int wait_for_last_wqe_event(entity &e) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:98]: wait_for_last_wqe_event_by_log

File: test/gtest/uct/ib/test_ib_event.cc [hunk 18]
```diff
@@ -65,38 +75,98 @@ class uct_p2p_test_event : public uct_p2p_test {
...
+        ASSERT_UCS_OK_OR_INPROGRESS(status);
+        if (status == UCS_OK) {
+            return 1;
+        }
+
+        if (!before) {
+            trigger_last_wqe_event(e);
+        }
+
+        deadline = ucs_get_time() +
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:152]: ucs_time_t deadline = ...

File: test/gtest/uct/ib/test_ib_event.cc [hunk 19]
```diff
@@ -65,38 +75,98 @@ class uct_p2p_test_event : public uct_p2p_test {
...
+                   ucs_time_from_sec(ucs::test_time_multiplier());
+        while (ucs_get_time() < deadline) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:154]: while ((ucs_get_time() < deadline) && !got_event) {
  progress();
}

return got_event

File: test/gtest/uct/ib/test_ib_event.cc [hunk 20]
```diff
@@ -115,6 +185,28 @@ UCS_TEST_P(uct_p2p_test_event, last_wqe, "ASYNC_EVENTS=y")
...
+    const p2p_resource *r = dynamic_cast<const p2p_resource*>(GetParam());
+    ucs_assert_always(r != NULL);
+
+    ASSERT_TRUE(wait_for_last_wqe_event_cb(sender(), false));
+    if (!r->loopback) {
+        ASSERT_TRUE(wait_for_last_wqe_event_cb(receiver(), false));
+    }
+}
+
+UCS_TEST_P(uct_p2p_test_event, last_wqe_cb_nosched)
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:199]: last_wqe_cb_before_subscribe

File: test/gtest/uct/ib/test_ib_event.cc [hunk 21]
```diff
@@ -115,6 +185,28 @@ UCS_TEST_P(uct_p2p_test_event, last_wqe, "ASYNC_EVENTS=y")
     }
 }
 
+UCS_TEST_P(uct_p2p_test_event, last_wqe_cb)
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_event.cc:188]: last_wqe_cb_after_subscribe

File: src/uct/ib/base/ib_device.c (hunk 3 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c:210]: event before subscription (i.e. before tx flushed)

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/base/ib_device.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/base/ib_device.c:210]: but then i think we would not have any entry at all?

File: src/uct/ib/base/ib_device.c [hunk 22]
```diff
@@ -176,6 +195,86 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+ucs_status_t
+uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
+                                    enum ibv_event_type event_type,
+                                    uint32_t resource_id,
+                                    ucs_callbackq_t *cbq,
+                                    ucs_callback_t cb, void *arg)
+{
+    uct_ib_async_event_t event;
+    uct_ib_async_event_val_t *entry;
+    ucs_status_t status = UCS_ERR_NO_MEMORY;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:234]: initalized var should be 1st

**Comment by yosefe** [comment]
@brminich @dmitrygx can you pls look as well?

File: src/uct/ib/base/ib_device.c (hunk 3 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c:210]: i mean event arrived before. flag=1, no subscription yet.

File: src/uct/ib/base/ib_device.c [hunk 23]
```diff
@@ -176,6 +195,86 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+        if (entry->cb != NULL) {
+            ucs_callbackq_add_safe(entry->cbq, entry->cb, entry->arg,
+                                   UCS_CALLBACKQ_FLAG_ONESHOT);
+            kh_del(uct_ib_async_event, &dev->async_events_hash, iter);
+        } else {
+            entry->flag = 1;
+        }
+    } else if ((ret == UCS_KH_PUT_BUCKET_EMPTY) ||
+               (ret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        entry->flag = 1;
```
Comment (brminich) [inline src/uct/ib/base/ib_device.c:219]: why need to put it and just ignore?

File: src/uct/ib/base/ib_device.c [hunk 24]
```diff
@@ -176,6 +195,76 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+static void
+uct_ib_device_async_event_dispatch(uct_ib_device_t *dev,
+                                   const uct_ib_async_event_t *event)
+{
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+    int ret;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter  = kh_put(uct_ib_async_event, &dev->async_events_hash, *event, &ret);
```
Comment (brminich) [inline src/uct/ib/base/ib_device.c:207]: can use get now, looks like no need for put

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c:207]: that was the purpose of the flag - it just looks like no need.
we do nothing if it's new entry, but we still add it so following put will return KEY_PRESENT

File: src/uct/ib/base/ib_device.c [hunk 25]
```diff
@@ -176,6 +195,79 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    entry = &kh_value(&dev->async_events_hash, iter);
```
Comment (brminich) [inline src/uct/ib/base/ib_device.c:208]: minor: this entry initialization can be moved to `if (ret == UCS_KH_PUT_KEY_PRESENT)`

File: src/uct/ib/base/ib_device.c [hunk 26]
```diff
@@ -176,6 +195,79 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+    int ret;
+
+    event.event_type  = event_type;
+    event.resource_id = resource_id;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter  = kh_put(uct_ib_async_event, &dev->async_events_hash, event, &ret);
+    entry = &kh_value(&dev->async_events_hash, iter);
```
Comment (brminich) [inline src/uct/ib/base/ib_device.c:238]: minor: this entry initialization can be moved to `else if (...`

**Review by brminich — APPROVED** [review]

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

File: src/uct/ib/base/ib_device.c [hunk 27]
```diff
@@ -52,6 +52,25 @@ KHASH_IMPL(uct_ib_ah, struct ibv_ah_attr, struct ibv_ah*, 1,
            uct_ib_kh_ah_hash_func, uct_ib_kh_ah_hash_equal)
 
 
+static UCS_F_ALWAYS_INLINE
+khint32_t uct_ib_async_event_hash_func(uct_ib_async_event_t event)
+{
+    return kh_int64_hash_func(((uint64_t)event.event_type << 32) |
+                              event.resource_id);
+}
+
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:61]: remove space line

File: src/uct/ib/base/ib_device.c [hunk 28]
```diff
@@ -176,6 +195,79 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_put(uct_ib_async_event, &dev->async_events_hash, *event, &ret);
+    if (ret == UCS_KH_PUT_KEY_PRESENT) {
+        entry = &kh_value(&dev->async_events_hash, iter);
+        ucs_callbackq_add_safe(entry->cbq, entry->cb, entry->arg,
+                               UCS_CALLBACKQ_FLAG_ONESHOT);
+        kh_del(uct_ib_async_event, &dev->async_events_hash, iter);
+    } else if ((ret == UCS_KH_PUT_BUCKET_EMPTY) ||
+               (ret == UCS_KH_PUT_BUCKET_CLEAR)) {
+        /* entry added */
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:215]: seems pointless if
maybe remove it and assert ret==EMPTY||ret==CLEAR

Comment (Artemy-Mellanox) [reply src/uct/ib/base/ib_device.c:215]: @brminich suggested it for readability

**Review by yosefe — APPROVED** [review]

File: src/uct/ib/base/ib_device.c [hunk 29]
```diff
@@ -176,6 +194,89 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+                                   const uct_ib_async_event_t *event)
+{
+    uct_ib_async_event_val_t *entry;
+    khiter_t iter;
+
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_get(uct_ib_async_event, &dev->async_events_hash, *event);
+    if (iter != kh_end(&dev->async_events_hash)) {
+        entry = &kh_value(&dev->async_events_hash, iter);
+        if (entry->cb) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:208]: entry->cb != NULL

File: src/uct/ib/base/ib_device.c [hunk 30]
```diff
@@ -176,6 +194,89 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+                                   UCS_CALLBACKQ_FLAG_ONESHOT);
+        } else {
+            entry->flag = 1;
+        }
+    }
+    ucs_spin_unlock(&dev->async_event_lock);
+}
+
+ucs_status_t
+uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:219]: 1. uct_ib_device_async_event_register (create entry with kh_put) - set cbq and resource_id
2. uct_ib_device_async_event_wait (type, res_id, cb, arg)

File: src/uct/ib/base/ib_device.c [hunk 31]
```diff
@@ -176,6 +194,89 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    entry = &kh_value(&dev->async_events_hash, iter);
+    if (ret == UCS_KH_PUT_KEY_PRESENT) {
+        if (entry->flag) {
+            status = UCS_OK;
+            goto out;
+        } else if (entry->cb) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:246]: == NULL

File: src/uct/ib/base/ib_device.c [hunk 32]
```diff
@@ -238,22 +237,48 @@ uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
...
-            status = UCS_ERR_BUSY;
-            goto out;
-        }
+
+    if (entry->flag) {
+        status = UCS_OK;
+        goto out;
+    } else if (entry->cb != NULL) {
+        status = UCS_ERR_BUSY;
+        goto out;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c]: we can replace goto out by else:
if (flag){
 status=OK
} else if (cb!=NULL) {
 status=BUSY
} else {
 entry->cb=cb;
 ...
}

File: src/uct/ib/base/ib_device.c [hunk 33]
```diff
@@ -176,6 +194,114 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    entry->cb  = cb;
+    entry->arg = arg;
+    status     = UCS_INPROGRESS;
+
+out:
+    ucs_spin_unlock(&dev->async_event_lock);
+    return status;
+}
+
+void uct_ib_device_async_event_clear(uct_ib_device_t *dev,
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:288]: uct_ib_device_async_event_unregister

File: src/uct/ib/base/ib_device.c [hunk 34]
```diff
@@ -457,6 +586,14 @@ void uct_ib_device_cleanup(uct_ib_device_t *dev)
 
     ucs_debug("destroying ib device %s", uct_ib_device_name(dev));
 
+    ucs_assert(kh_size(&dev->async_events_hash) == 0);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:589]: ucs_warn

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Artemy-Mellanox pls squash

**Pull request merged**